tests/eas/generic.py

```diff
@@ -0,0 +1,433 @@
+        :meth:EnergyModel.estimate_from_cpu_util to get a DataFrame showing the
+        estimated power usage over time under ideal EAS behaviour.
+
+        :param experiment: The :class:Experiment to examine
+        :returns: A Pandas DataFrame with a column each node in the energy model
+                  (keyed with a tuple of the CPUs contained by that node) and a
+                  "power" column with the sum of other columns. Shows the
+                  estimated *optimal* power over time.
+        """
+        task_utils_df = self.get_task_utils_df(experiment)
+
+        nrg_model = self.te.nrg_model
+
+        def exp_power(row):
+            task_utils = row.to_dict()
+            expected_utils = nrg_model.get_optimal_placements(task_utils)
+            power = nrg_model.estimate_from_cpu_util(expected_utils[0])
+            columns = power.keys()
+            return pd.Series([power[c] for c in columns], index=columns)
+        return task_utils_df.apply(exp_power, axis=1)
+
+    def _test_slack(self, experiment, tasks):
+        """
+        Assert that the RTApp workload was given enough performance
+
+        Use :class:PerfAnalysis to find instances where the experiment's RT-App
+        workload wasn't able to complete its activations (i.e. its reported
+        "slack" was negative). Assert that this happened less that
+        ``negative_slack_allowed_pct`` percent of the time.
+
+        :meth:_test_task_placement asserts that estimated energy usage was
+        low. That will pass for runs where too *little* energy was used,
+        compromising performance. This method provides a separate test to
+        counteract that problem.
+        """
+
+        pa = PerfAnalysis(experiment.out_dir)
+        for task in tasks:
+            slack = pa.df(task)["Slack"]
+
+            bad_activations_pct = len(slack[slack < 0]) * 100. / len(slack)
+            if bad_activations_pct > self.negative_slack_allowed_pct:
+                raise AssertionError("task {} missed {}% of activations".format(
+                    task, bad_activations_pct))
+
+    def _test_task_placement(self, experiment, tasks):
+        """
+        Test that task placement was energy-efficient
+
+        Use :meth:get_expected_power_df and :meth:get_power_df to estimate
+        optimal and observed power usage for task placements of the experiment's
+        workload. Assert that the observed power does not exceed the optimal
+        power by more than 20%.
+        """
+        exp_power = self.get_expected_power_df(experiment)
+        est_power = self.get_power_df(experiment)
+
+        exp_energy = area_under_curve(exp_power['power'])
+        est_energy = area_under_curve(est_power['power'])
+
+        msg = 'Estimated {} bogo-Joules to run workload, expected {}'.format(
+            est_energy, exp_energy)
+        self.assertLess(est_energy, exp_energy * 1.2, msg=msg)
+
+class OneSmallTask(EnergyModelTest):
```

Comment:
I thought generic meant more that it would hold the base EnergyModelTest class, which then can be extended in specific test classes. I find it more clear to have the EnergyModelTest class in its own file but it's just a preference.